openai/src/commonMain/kotlin/com/xebia/functional/xef/auto/llm/openai/OpenAI.kt

```diff
@@ -3,39 +3,43 @@ package com.xebia.functional.xef.auto.llm.openai
 import com.xebia.functional.xef.env.getenv
 import kotlin.jvm.JvmField
 
-class OpenAI(internal val token: String) {
-  val GPT_4 = OpenAIModel(this, "gpt-4", ModelType.GPT_4)
+class OpenAI(internal val token: String) : AutoCloseable, AutoClose by autoClose() {
+  val GPT_4 by lazy { autoClose(OpenAIModel(this, "gpt-4", ModelType.GPT_4)) }
 
-  val GPT_4_0314 = OpenAIModel(this, "gpt-4-0314", ModelType.GPT_4)
+  val GPT_4_0314 by lazy { autoClose(OpenAIModel(this, "gpt-4-0314", ModelType.GPT_4)) }
 
-  val GPT_4_32K = OpenAIModel(this, "gpt-4-32k", ModelType.GPT_4_32K)
+  val GPT_4_32K by lazy { autoClose(OpenAIModel(this, "gpt-4-32k", ModelType.GPT_4_32K)) }
 
-  val GPT_3_5_TURBO = OpenAIModel(this, "gpt-3.5-turbo", ModelType.GPT_3_5_TURBO)
+  val GPT_3_5_TURBO by lazy { autoClose(OpenAIModel(this, "gpt-3.5-turbo", ModelType.GPT_3_5_TURBO)) }
 
-  val GPT_3_5_TURBO_16K = OpenAIModel(this, "gpt-3.5-turbo-16k", ModelType.GPT_3_5_TURBO_16_K)
+  val GPT_3_5_TURBO_16K by lazy { autoClose(OpenAIModel(this, "gpt-3.5-turbo-16k", ModelType.GPT_3_5_TURBO_16_K)) }
 
-  val GPT_3_5_TURBO_FUNCTIONS =
-    OpenAIModel(this, "gpt-3.5-turbo-0613", ModelType.GPT_3_5_TURBO_FUNCTIONS)
+  val GPT_3_5_TURBO_FUNCTIONS by lazy {
+    autoClose(OpenAIModel(this, "gpt-3.5-turbo-0613", ModelType.GPT_3_5_TURBO_FUNCTIONS))
+  }
 
-  val GPT_3_5_TURBO_0301 = OpenAIModel(this, "gpt-3.5-turbo-0301", ModelType.GPT_3_5_TURBO)
+  val GPT_3_5_TURBO_0301 by lazy { autoClose(OpenAIModel(this, "gpt-3.5-turbo-0301", ModelType.GPT_3_5_TURBO)) }
 
-  val TEXT_DAVINCI_003 = OpenAIModel(this, "text-davinci-003", ModelType.TEXT_DAVINCI_003)
+  val TEXT_DAVINCI_003 by lazy { autoClose(OpenAIModel(this, "text-davinci-003", ModelType.TEXT_DAVINCI_003)) }
 
-  val TEXT_DAVINCI_002 = OpenAIModel(this, "text-davinci-002", ModelType.TEXT_DAVINCI_002)
+  val TEXT_DAVINCI_002 by lazy { autoClose(OpenAIModel(this, "text-davinci-002", ModelType.TEXT_DAVINCI_002)) }
 
-  val TEXT_CURIE_001 = OpenAIModel(this, "text-curie-001", ModelType.TEXT_SIMILARITY_CURIE_001)
+  val TEXT_CURIE_001 by lazy { autoClose(OpenAIModel(this, "text-curie-001", ModelType.TEXT_SIMILARITY_CURIE_001)) }
 
-  val TEXT_BABBAGE_001 = OpenAIModel(this, "text-babbage-001", ModelType.TEXT_BABBAGE_001)
+  val TEXT_BABBAGE_001 by lazy { autoClose(OpenAIModel(this, "text-babbage-001", ModelType.TEXT_BABBAGE_001)) }
 
-  val TEXT_ADA_001 = OpenAIModel(this, "text-ada-001", ModelType.TEXT_ADA_001)
+  val TEXT_ADA_001 by lazy { autoClose(OpenAIModel(this, "text-ada-001", ModelType.TEXT_ADA_001)) }
 
-  val TEXT_EMBEDDING_ADA_002 =
-    OpenAIModel(this, "text-embedding-ada-002", ModelType.TEXT_EMBEDDING_ADA_002)
+  val TEXT_EMBEDDING_ADA_002 by lazy {
+    autoClose(OpenAIModel(this, "text-embedding-ada-002", ModelType.TEXT_EMBEDDING_ADA_002))
+  }
 
-  val DALLE_2 = OpenAIModel(this, "dalle-2", ModelType.GPT_3_5_TURBO)
+  val DALLE_2 by lazy { autoClose(OpenAIModel(this, "dalle-2", ModelType.GPT_3_5_TURBO)) }
```

Comment:
The issue with autoclosing resources declared in the companion like this, whose lifetime is the entire application and not initialized with a function, is that once you close them, they will become unusable to be reopened since the companion state will point to a stale object. You'd need to change all these values to function or perform the Autoclosing on JVM shutdown hooks that await termination.

Maybe I misunderstood `autoClose,` and this is safe, but when you run this through a scope, once the Scope ends, this object will become stale for the remainder of the application.

In practice, the resources this object holds are just the underlying OpenAI or similar HTTP clients, which don't do persistence or anything relevant on shutdown. \do they need to be closed and created on each Scope?, or can we have these objects for the entire Application lifecycle ? //cc @nomisRev

Reply:
> The issue with autoclosing resources declared in the companion like this

These are defined in the `OpenAI` class not the `companion object`? 🤔 I think you understood `autoClose` correct, it's basically the same as `autoCloseable { }`, or `install`, inside `ResourceScope`.

Your rationale applies to the `DEFAULT`, `DEFAULT_CHAT`, etc instances defined below in the `companion object`  which creates instances of the `OpenAI` class holding these other instances. The `DEAULT_X` instances are currently still leaking. Which are currently not _closing_ only any scope and thus leaking. It's currently not scoped to the application scope, or the `AIScope` in any way. Which I think would be ideal, but some changes in the entire `xef` library seem to be needed for that.

> which don't do persistence or anything relevant on shutdown. \do they need to be closed and created on each Scope?

I would say yes, because they are defined as `Closeable` themselves. Typically they will hold some `ExecutorService`, or `CoroutineScope`, and often a bunch of other stateful things inside. Especially when composing different blocks together, which might have different scopes I'd say it become more important. In small examples that, or single scope applications, you won't face any issues.

Another issue is that currently we're hardcoding to `openAITokenFromEnv` given that these instances are singletons, and thus we cannot provide custom configurations. This is a roadblock for web for example.

Reply:
Makes sense @nomisRev resolving this and we can address those issues in separate PRs